accepted/future-releases/static-extension-methods/design-document.md

```diff
@@ -0,0 +1,504 @@
+
+The methods we allow you to add this way are normal methods, operators, getter and setters.
+
+## Declaring Static Extension Methods
+
+### Syntax
+
+A static extension of a type is declared using syntax like:
+
+```dart
+extension MyFancyList<T> on List<T> {
+  int get doubleLength => this.length * 2;
+  List<T> operator-() => this.reversed.toList();
+  List<List<T>> split(int at) => 
+      <List<T>>[this.sublist(0, at), this.sublist(at)];
+  List<T> mapToList<R>(R Function(T) convert) => this.map(convert).toList();
+}
+```
+
+More precisely, an extension declaration is a top-level declaration with a grammar similar to:
+
+```ebnf
+<extension> ::= 
+  `extension' <identifier>? <typeParameters>? `on' <type> `?'? `{'
+     memberDeclaration*
+  `}'
+```
+
+Such a declaration introduces its *name* (the identifier) into the surrounding scope. The name does not denote a type, but it can be used to denote the extension itself in various places. The name can be hidden or shown in `import` or `export` declarations.
+
+The *type* can be any valid Dart type, including a single type variable. It can refer to the type parameters of the extension. It can be followed by `?` which means that it allows `null` values. When Dart gets non-nullable types by default (NNBD), this `?` syntax is removed and subsumed by nullable types like `int?` being allowed in the `<type>` position.
+
+The member declarations can be any non-abstract static or instance member declaration except for instance variables and constructors. Abstract members are not allowed since the extension declaration does not introduce an interface, and constructors are not allowed because the extension declaration doesn't introduce any type that can be constructed. Instance variables are not allowed because there won't be any memory allocation per instance that the extension applies to. We could implement instance variables using an `Expando`, but it would necessarily be nullable, so it would still not be an actual instance variable.
+
+### Omitting Names For Local Extensions
+
+If an extension declaration is only used locally in a library, there is no need to worry about naming conflicts or overrides. In that case, then name identifier can be omitted (hence the `<identifier>?` in the grammar above).
+
+Example:
+
+```dart
+extension<T> on List<T> {
+  void quadruple() { ... }
+}
+```
+
+This is equivalent to giving the extension a fresh private name.
+
+We may need to make `on` a built-in identifier, and not allow those as names of extensions, then there should not be any parsing issue. Even without that, the grammar should be unambiguous because `extension on on on { … }` and `extension on on { … }` are distinguishable, and the final type cannot be empty. It may be *harder* to parse, though.
+
+This is a simple feature, but with very low impact. It only allows you to omit a single private name for an extension that is only used in a single library.
+
+### Scope
```

Comment:
Should we provide a way to show/hide individual members of an extension?  Otherwise you're somewhat stuck if you have two extensions that provide overlapping names and you want to pick and choose.